### PR TITLE
Collaborators

### LINKED ISSUE #37: Implement collaborators

### LINKED ISSUE #60: Checks for collaborators for getting the user authorizations
- [x]  In collaborators addition exculde the owner him self to add into collaborators
- [x] A collaborator can do the following for a repo 
  - [x] he can pin issue and unpin issue
  - [x] he can lock/unlock the issue 
  - [x] he can open and close the issue        

Unrelated:

  - [ ]  can open any issue in the collaborated repo (for private repo only)
  - [ ] he can write a new issue (for private repo only)
  - [ ] he can post comments in any issue of the collaborated repo (for private repo only)

### PR DESCRIPTION
**Pull request by sridhar02**
Fixes #37 
Fixes #60

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -749,25 +749,38 @@ func getCollaboratorPageHandler(c *gin.Context, db *sql.DB) {
...
 		}
-		collaborator = Collaborator{
+		collaborator := Collaborator{
 			Username:  UserName,
 			UserImage: image,
 		}
+		collaborators = append(collaborators, collaborator)
+	}
+
+	IsCollaboratorsAvailable := false
```
Comment (vramana) [inline main.go]: `IsCollaboratorsAvailable :=  len(collaborators) >= 1`

File: main.go [hunk 2]
```diff
@@ -782,9 +795,20 @@ func postCollaboratorPageHandler(c *gin.Context, db *sql.DB) {
...
 
 	repoId := c.PostForm("repo_id")
-	userId := c.PostForm("user_id")
 
-	_, err = db.Exec(`INSERT INTO collaborators(repo_id,user_id)VALUES($1,$2,)`, repoId, userId)
+	userName := c.PostForm("user_name")
+
+	var userId string
+
+	row := db.QueryRow(`SELECT id from users WHERE username = $1`, userName)
```
Comment (vramana) [inline main.go]: Instead of custom query use GetUserByUsername function we already wrote.

File: templates/collaboration.html [hunk 3]
```diff
@@ -24,37 +24,82 @@
...
 		<span>Push access to the repository</span>
 	</div>
-	<div class="info">
+	{{if .IsCollaboratorsAvailable}}
+		{{range .Collaborators}}
+		<div class="collaborator">
+			<div class="user-photo"><img class="user-photo" src="{{.UserImage}}"></div>
+			<div class="username"><a href="#">{{.Username}}</a></div>
+		</div>	
+	{{end}}
```
Comment (vramana) [inline templates/collaboration.html]: This end is for range block. Indentation is wrong.

File: main.go [hunk 4]
```diff
@@ -782,9 +795,20 @@ func postCollaboratorPageHandler(c *gin.Context, db *sql.DB) {
 	}
 
 	repoId := c.PostForm("repo_id")
```
Comment (vramana) [inline main.go]: remove this field. Instead of getting it from form. Derive it from repo_name param (get current repo).

File: main.go [hunk 5]
```diff
@@ -714,10 +714,10 @@ type Collaborator struct {
 
 func getCollaboratorPageHandler(c *gin.Context, db *sql.DB) {
 
-	currentUser, err := authorize(c, db)
-	if err == nil {
-		c.Redirect(http.StatusFound, "http://localhost:8000/"+currentUser.Username)
-		return
+	_, err := authorize(c, db)
+	authorized := true
```
Comment (vramana) [inline main.go]: `authorized := err != nil`

File: main.go [hunk 6]
```diff
@@ -707,6 +707,117 @@ func postUnlockPageHandler(c *gin.Context, db *sql.DB) {
...
 }
 
+type Collaborator struct {
+	Username  string
+	UserImage string
+}
+
+func getCollaboratorPageHandler(c *gin.Context, db *sql.DB) {
+
+	_, err := authorize(c, db)
```
Comment (vramana) [inline main.go]: Collaborator page should be only visible to repo owner.

File: main.go (hunk 1 above)
Comment (sridhar02) [reply main.go]: DONE

File: main.go (hunk 2 above)
Comment (sridhar02) [reply main.go]: DONE

File: templates/collaboration.html (hunk 3 above)
Comment (sridhar02) [reply templates/collaboration.html]: DONE

File: main.go (hunk 4 above)
Comment (sridhar02) [reply main.go]: DONE

File: main.go (hunk 6 above)
Comment (sridhar02) [reply main.go]: Done

File: main.go (hunk 5 above)
Comment (sridhar02) [reply main.go]: Done

File: main.go [hunk 7]
```diff
@@ -707,6 +707,163 @@ func postUnlockPageHandler(c *gin.Context, db *sql.DB) {
...
+	}
+
+	currentRepo, err := getCurrentRepo(db, username, repoName)
+	if err != nil {
+		fmt.Println(err)
+		c.AbortWithStatus(http.StatusInternalServerError)
+		return
+	}
+
+	rows, err := db.Query(`SELECT users.username,users.image,users.name FROM users JOIN collaborators ON users.id = collaborators.user_id
```
Comment (vramana) [inline main.go]: format the query a bit nicely into multiple lines.

File: main.go [hunk 8]
```diff
@@ -707,6 +707,163 @@ func postUnlockPageHandler(c *gin.Context, db *sql.DB) {
...
+								WHERE collaborators.repo_id = $1;`, currentRepo.RepoId)
+	if err != nil {
+		fmt.Println(err)
+		c.AbortWithStatus(http.StatusInternalServerError)
+		return
+	}
+
+	collaborators := []Collaborator{}
+
+	var UserName, image, name string
```
Comment (vramana) [inline main.go]: move variable declaration inside the for loop. Since they are not used outside elsewhere.

File: main.go [hunk 9]
```diff
@@ -707,6 +707,163 @@ func postUnlockPageHandler(c *gin.Context, db *sql.DB) {
...
+			"IsCollaboratorsAvailable": IsCollaboratorsAvailable,
+		})
+
+}
+
+func postCollaboratorPageHandler(c *gin.Context, db *sql.DB) {
+
+	currentUser, err := authorize(c, db)
+	if err != nil {
+		c.Redirect(http.StatusFound, "http://localhost:8000/"+currentUser.Username)
```
Comment (vramana) [inline main.go]: This is wrong. If user is not authorized, then currentUser object is an empty struct.  It should redirect to login page.

File: main.go [hunk 10]
```diff
@@ -707,6 +707,163 @@ func postUnlockPageHandler(c *gin.Context, db *sql.DB) {
...
+	userName := c.PostForm("user_name")
+
+	user, err := GetUserByUserName(db, userName)
+	if err != nil {
+		fmt.Println(err)
+		c.AbortWithStatus(http.StatusInternalServerError)
+		return
+	}
+
+	_, err = db.Exec(`INSERT INTO collaborators(repo_id,user_id)VALUES($1,$2)`, currentRepo.RepoId, user.ID)
```
Comment (vramana) [inline main.go]: Prettify query 

```
INSERT INTO collaborators(repo_id, user_id) VALUES ($1,$2)
```

File: main.go [hunk 11]
```diff
@@ -707,6 +707,163 @@ func postUnlockPageHandler(c *gin.Context, db *sql.DB) {
...
+
+func postRemoveCollaboratorHandler(c *gin.Context, db *sql.DB) {
+
+	currentUser, err := authorize(c, db)
+	if err != nil {
+		c.Redirect(http.StatusFound, "http://localhost:8000/"+currentUser.Username)
+		return
+	}
+
+	fmt.Println("error")
```
Comment (vramana) [inline main.go]: Remove unnecessary fmt.Println

File: main.go [hunk 12]
```diff
@@ -707,6 +707,163 @@ func postUnlockPageHandler(c *gin.Context, db *sql.DB) {
...
+	userName := c.PostForm("user_name")
+
+	user, err := GetUserByUserName(db, userName)
+	if err != nil {
+		fmt.Println(err)
+		c.AbortWithStatus(http.StatusInternalServerError)
+		return
+	}
+
+	fmt.Println("userId")
```
Comment (vramana) [inline main.go]: Remove unnecessary fmt.Println

File: templates/collaboration.html [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+}
+
+.collaborator-details{
+	display: flex;
+}
+
+
+</style>
+{{template "nav.html" .}}
+{{if .Authorized}}
```
Comment (vramana) [inline templates/collaboration.html]: No need for two if statements. We can use `and` function like `eq`. Write 

```
{{ if and .Authorized .IsRepoOwner }}
     ....
{{ end }}
````

File: main.go (hunk 8 above)
Comment (sridhar02) [reply main.go]: Done

File: main.go (hunk 9 above)
Comment (sridhar02) [reply main.go]: done

File: main.go (hunk 10 above)
Comment (sridhar02) [reply main.go]: done

File: templates/collaboration.html (hunk 13 above)
Comment (sridhar02) [reply templates/collaboration.html]: Done

File: main.go [hunk 14]
```diff
@@ -272,8 +272,8 @@ func getIssuePageHandler(c *gin.Context, db *sql.DB) {
 
 	locked := lock == "Locked" && IsRepoOwner == false
 
-	IsRepoOwnerOrIssueUser := currentUser.Username == c.Param("user_name") ||
-		currentUser.Username == issue.Username
+	// IsRepoOwnerOrIssueUser := currentUser.Username == c.Param("user_name") ||
```
Comment (vramana) [inline main.go]: Remove.

File: main.go [hunk 15]
```diff
@@ -314,6 +316,24 @@ func createIssueComment(c *gin.Context, db *sql.DB) {
...
+		return
+	}
+
+	writeAccess, err := hasWriteAccess(db, currentRepo, currentUser)
+	if err != nil {
+		fmt.Println(err)
+		c.AbortWithStatus(http.StatusInternalServerError)
+		return
+	}
+	if !writeAccess {
```
Comment (vramana) [inline main.go]: Why did we add this here at all ? This is create issue comment any authorized user should be able to comment.

Comment (sridhar02) [reply main.go]: Done

File: main.go (hunk 14 above)
Comment (sridhar02) [reply main.go]: DONE

**Review by vramana — APPROVED** [review]

**Pull request merged**